param/parameterized.py

```diff
@@ -1139,36 +1143,69 @@ def __repr__(self):
 
 
 
-
-    def script_repr(self,imports=[],prefix="    "):
+    def script_repr(self, imports=[], prefix=" ", unknown_value='<?>', qualify=False):
         """
         Variant of __repr__ designed for generating a runnable script.
         """
-        # Suppresses automatically generated names.
-        settings=[]
-        for name,val in self.get_param_values(onlychanged=script_repr_suppress_defaults):
-            if name == 'name' and (val is not None and
-                                   re.match('^'+self.__class__.__name__+'[0-9]+$',val)):
-                rep=None
-            else:
-                rep=script_repr(val,imports,prefix,settings)
-
-            if rep is not None:
-                settings.append('%s=%s' % (name,rep))
-
+        exclude=['self', 'name']
 
         # Generate import statement
         mod = self.__module__
-
         bits = mod.split('.')
-
         imports.append("import %s"%mod)
         imports.append("import %s"%bits[0])
 
-        # CB: Doesn't give a nice repr, but I don't see what to do
-        # otherwise that will work in all cases. Also I haven't
-        # updated this code in other places (e.g. simulation).
-        return mod+'.'+self.__class__.__name__ + "(" + (",\n"+prefix).join(settings) + ")"
+        changed_params = dict(self.get_param_values(onlychanged=script_repr_suppress_defaults))
+        param_values = dict(self.get_param_values())
+        spec = inspect.getargspec(self.__init__)
+        args = spec.args[1:] if spec.args[0] == 'self' else spec.args
+
+        if spec.defaults is not None:
+            posargs = spec.args[:-len(spec.defaults)]
+            kwargs = dict(zip(spec.args[-len(spec.defaults):], spec.defaults))
+        else:
+            posargs, kwargs = args, []
+
+        ordering = sorted(
+            sorted(changed_params.keys()), # alphanumeric tie-breaker
+            key=lambda k: (- float('inf')  # No precedence is lowest possible precendence
+                           if self.params(k).precedence is None
+                           else self.params(k).precedence))
+
+        values = dict(self.get_param_values())
+
+        arglist, keywords, processed = [], [], []
+        for k in args + ordering:
+            if k in processed: continue
+
+            # Suppresses automatically generated names.
+            if k == 'name' and (values[k] is not None and
+                                re.match('^'+self.__class__.__name__+'[0-9]+$', values[k])):
+                continue
+
+            value = script_repr(values[k], imports, prefix=prefix,settings=[],
+                                unknown_value=unknown_value,
+                                qualify=qualify) if k in values else unknown_value
+            if value is None:
+                raise Exception("Argument %r is not a parameter "
+                                "and has an unknown value." % k)
+
```

Comment:
Could you clarify this part? It seems to be saying that if ``script_repr(values[k])`` returns ``None``, that means ``k`` is not a parameter of ``self``? I don't understand why getting back ``None`` from ``script_repr`` means ``k`` isn't a parameter. Anyway, I do agree that getting back ``None`` from ``script_repr`` means ``k``'s value is "unknown" (if that means, "has unknown repr")...but I don't agree about raising an exception. Param has ``script_repr_reg``, a place to register handlers, and I can see some of the ones param itself supplies actually return ``None``, e.g.

```python
def empty_script_repr(*args):
    return None

script_repr_reg[random.Random] = empty_script_repr
```

So, I could alter your code above to be ``if value is None: continue``, and I think that would match the previous behavior of ``script_repr`` (untested). However, I think maybe we should change that behavior. We could give a warning whenever we deliberately suppress a ``script_repr``, because people would probably like to be aware of what's being missed from their ``script_repr`` (e.g. random state). This would change the previous behavior of param (which was silently suppressing), but I think I would prefer it. What do you think?

If we did the above, I'm not sure if we'd need to distinguish deliberate suppression (i.e. a deliberate ``None`` coming back from a handler) from the case of the ``script_repr`` function falling all the way through to ``repr`` and still getting back ``None``, because I'm not sure what kind of ``__repr__`` method would return ``None``. Basically, I'm not clear about when you thought ``value`` would be ``None``, above...